### PR TITLE
Client proof types

### PR DESCRIPTION
**Pull request by dtebbs**
Stricter types for generic Proof and VerificationKey types and snark API refactor.
(Allows more versatility in data structures that use proofs, verification keys and extended proofs)

### CODE REVIEW

File: client/test_commands/scenario.py [hunk 1]
```diff
@@ -234,11 +236,11 @@ def compute_h_sig_attack_nf(
     assert attack_primary_input3 != 0
     assert attack_primary_input4 != 0
 
-    print("proof_json => ", proof_json)
-    print("proof_json[inputs][3] => ", proof_json["inputs"][3])
-    print("proof_json[inputs][4] => ", proof_json["inputs"][4])
-    proof_json["inputs"][3] = hex(attack_primary_input3)
-    proof_json["inputs"][4] = hex(attack_primary_input4)
+    print("proof => ", proof)
+    print("proof[inputs][3] => ", proof.inputs[3])
```
Comment (AntoineRondelet) [inline client/test_commands/scenario.py:240]: Why not: `print("proof.inputs[3] = ", proof.inputs[3])`? (same for the line below)

File: client/tests/test_zksnark.py [hunk 2]
```diff
@@ -2,18 +2,18 @@
...
 from unittest import TestCase
 
 
 # pylint: disable=line-too-long
-VERIFICATION_KEY_BLS12_377_GROTH16 = {
+VERIFICATION_KEY_BLS12_377_GROTH16 = Groth16.VerificationKey.from_json_dict({
     "alpha": [
         "0x009d7309d79d5215384a7a9a1d9372af909582781f388a51cb833c87b8024519cf5b343cb35d49a5aa52940f14b7b8e7",  # noqa
-        "0x012816ef6069ef1e40eaab0a111f9b98b276dbf2a3209d788eb8ce635ce92a29c2bcdaa3bb9b375a8d3ee4325c07f4ea"  # noqa
+        "0x012816ef6069ef1e40eaab0a111f9b98b276dbf2a3209d788eb8ce635ce92a29c2bcdaa3bb9b375a8d3ee4325c07f4ea",  # noqa
```
Comment (AntoineRondelet) [inline client/tests/test_zksnark.py:16]: Why did you need to add this comma here?

File: client/test_commands/scenario.py (hunk 1 above)
Comment (dtebbs) [reply client/test_commands/scenario.py]: The original code predates me, so not sure why it was `=>` instead of `=`.  Changed to `=`.

File: client/tests/test_zksnark.py (hunk 2 above)
Comment (dtebbs) [reply client/tests/test_zksnark.py:16]: It's not needed as such.  Many of our lists and dictionaries have a trailing comma, so I added it to match those.

File: client/test_commands/scenario.py (hunk 1 above)
Comment (AntoineRondelet) [reply client/test_commands/scenario.py]: Oh yes, there were two things in my comment above:
- `=>` -> `=`
- `proof.[inputs][3]` -> `proof.inputs[3]`

File: client/tests/test_zksnark.py (hunk 2 above)
Comment (AntoineRondelet) [reply client/tests/test_zksnark.py:16]: Adding this comma here contrasts with the rest of the block actually: https://github.com/clearmatics/zeth/blob/client-proof-types/client/tests/test_zksnark.py#L13-L39
which itself contrasts with: https://github.com/clearmatics/zeth/blob/client-proof-types/client/tests/test_zksnark.py#L172-L198 (where you add final commas everywhere)

File: client/test_commands/scenario.py (hunk 1 above)
Comment (dtebbs) [reply client/test_commands/scenario.py]: updated

File: client/tests/test_zksnark.py (hunk 2 above)
Comment (dtebbs) [reply client/tests/test_zksnark.py]: updated.  (yes, there are examples of each - some data was copy-pasted from json)

**Comment by AntoineRondelet** [comment]
LGTM
